### DISCUSSION TITLE
Manual input for inertia matrix

### DISCUSSION THREAD

**Discussion by JorgenStarStruck** [capytaine/capytaine#321] [Q&A]
Hi, 

I am currently modelling a tripod floating wind turbine as a student hydraulic engineering. Since I want to avoid tedious computations, i use an stl mesh of the 3 cylinders only. To account for the presence of the connecting structure and the turbine on one of the floaters i want to add a force in meshmagick at the correct center of gravity and manually insert the inertia matrix. I manage to insert the gravitational force but i cant seem to figure out what inertia matrix Capytaine uses. It has a 6x6 matrix but can anyone help me with what inertia i have to use for what position in the tensor? I know i should use tha rotational inertia for the 3 by 3 in the right lower corner, but have no idea what is in the rest of the matrix. (Ixx = m*(z^2+y^2)). Please correct me if i am wrong or if you have any suggestions.

**Comment by mancellin** [comment]
If the rotation dofs are defined as rotations around the center of mass, then the mass matrix of a rigid body reads:

$$
M = \begin{pmatrix}
m & 0 & 0 & 0 & 0 & 0 \\
0 & m & 0 & 0 & 0 & 0 \\
0 & 0 & m & 0 & 0 & 0 \\
0 & 0 & 0 & I_{xx} & I_{xy} & I_{xz} \\
0 & 0 & 0 & I_{yx} & I_{yy} & I_{yz} \\
0 & 0 & 0 & I_{zx} & I_{zy} & I_{zz}
\end{pmatrix}
$$

where $m$ is the mass and $I$ is the rotational inertia matrix.

If the rotation dofs are not defined around the center of mass, there are additionnal off-diagonal terms depending on the distance between the center of rotation and the center of mass.

See for instance https://www.fossen.biz/wiley/pdf/Ch3.pdf (first reference I found, there might exist better ones).

**Reply by JorgenStarStruck** [reply]
![image](https://user-images.githubusercontent.com/130484908/231466969-30e7b57c-7b21-4b3e-88ca-fff8f47b170c.png)
When i use the upper code, meshmagick tells me the cylinders that are in the mesh weigh 0.007 tons. I wanted to to compute the inertiamatrix for the cylinders only, but this gives the following inertia matrix. So the first 3 elements you propose on the diagonal do not represent the mass since this should be 7kg and not 63. The first inertia matrix is the one given by meshmagick, but I have relocated the center of gravity to that of the whole structure and not only to that of the cylinders.
![image](https://user-images.githubusercontent.com/130484908/231467970-2fcfdd2e-ab9b-4344-b4a7-909d5d3905b1.png)

**Reply by mancellin** [reply]
From the screenshot you shared, it seems that Meshmagick is assuming the body is an empty shell of a given thickness. Capytaine computes the mass by assuming a full body with homogeneous density $\rho$ equal to 1000 kg/m³ by default (see code sample below). This might indeed deserve to be better documented.

Besides that, your inertia matrix is showing off-diagonal terms despite the rotation center being the same as the center of mass. It could be due to a bug in v1.4 that should have been fixed (#196) in the latest version.

```python
import numpy as np
import capytaine as cpt

length = 2.0
radius = 1.0
rho = 1000.0

c1 = cpt.mesh_vertical_cylinder(center=(0, 0, -1), length=length, radius=radius, resolution=(8, 24, 16))
c2 = c1.translated([2.0, 2.0, 0])
c3 = c1.translated([2.0, -2.0, 0])
cylinders_mesh = c1 + c2 + c3

body = cpt.FloatingBody(mesh=cylinders_mesh,
                        center_of_mass=(1.0, 1.0, 0.0),
                        dofs=cpt.rigid_body_dofs(rotation_center=(1.0, 1.0, 0.0)))
M = body.compute_rigid_body_inertia(rho=rho).values

print(M[0, 0], 3*length*np.pi*radius**2*rho)
# 18634.971247381498 18849.55592153876

with np.printoptions(precision=1):
    print(M)
# [[ 18635.       0.       0.       0.       0.      -0. ]
#  [     0.   18635.       0.      -0.       0.       0. ]
#  [     0.       0.   18635.       0.      -0.       0. ]
#  [     0.      -0.       0.   97734.1   6211.7   6211.7]
#  [     0.       0.      -0.    6211.7  48040.9 -18635. ]
#  [    -0.       0.       0.    6211.7 -18635.   96114.1]]
```

**Reply by JorgenStarStruck** [reply]
Is their a way to tell capytaine it is a shell with this thickness and 1450kg/m^3, the cylinders represent 3 pvc hollow floaters

**Reply by mancellin** [reply]
You can set a `mass` for the body and it will be used for the inertia matrix

```python
total_area = np.sum(cylinders_mesh.faces_areas)
thickness = 3e-3
density = 1450
body = cpt.FloatingBody(mesh=cylinders_mesh,
                        center_of_mass=(1.0, 1.0, 0.0),
                        dofs=cpt.rigid_body_dofs(rotation_center=(1.0, 1.0, 0.0)),
                        mass=total_area*thickness*density)
```

but the rotation inertia might still be the one of a filled body.

The rigid-body inertia computation in Capytaine is a bit limited. It is meant for quick testing, but you should refer to another tool if you need to model a non-trivial mass distribution.
